### PR TITLE
Support self upgrade

### PR DESCRIPTION
**Pull request by yanecc**
I have implemented the Windows part based on @ankit-pn's work at #38 and resolved the conflicts, it is now able to work as expected, namely upgrade to the latest version 0.4.1.

### CODE REVIEW

**Comment by aooohan** [comment]
> Is it possible to introduce some third-party libraries or refer to the implementation of third-party libraries to complete this function?

I don't like to introduce too many third party packages.

**Comment by jan-bar** [comment]
> I don't like to introduce too many third party packages.

can refer to related implementations
just wanted to get some inspiration from these libraries

**Comment by aooohan** [comment]
There is an issue. We need to ask for administrator privileges when upgrading for `vfox` installed by setup, which is currently not handled by this change.

I've fixed the issue, but I don't have write access to your repository, so I can't commit at the moment.

**Comment by yanecc** [comment]
> There is an issue. We need to ask for administrator privileges when upgrading for `vfox` installed by setup, which is currently not handled by this change.
> 
> I've fixed the issue, but I don't have write access to your repository, so I can't commit at the moment.

I have invited you as collaborator.

**Review by aooohan — APPROVED** [review]
I have tested it on Windows.

File: cmd/commands/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	}
+	exeDir, exeName := filepath.Split(exePath)
+
+	binURL, diffURL := generateUrls(currVersion, latestVersion)
+	tempFile := "vfox_latest.tar.gz"
+	if runtime.GOOS == "windows" {
+		tempFile = "vfox_latest.zip"
+	}
+	tempFile = filepath.Join(os.TempDir(), tempFile)
+	tempDir := filepath.Join(os.TempDir(), "vfox_upgrade")
```
Comment (Chance-fyi) [inline cmd/commands/upgrade.go:139]: Use `VFOX_HOME/tmp`
See #156

Comment (Chance-fyi) [reply cmd/commands/upgrade.go:139]: `VFOX_HOME` may also be on a different file system.

Comment (yanecc) [reply cmd/commands/upgrade.go]: All right.

File: cmd/commands/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+	exeDir, exeName := filepath.Split(exePath)
+	binURL, diffURL := generateUrls(currVersion, latestVersion)
+	tempFile := "vfox_latest.tar.gz"
+	if runtime.GOOS == "windows" {
+		tempFile = "vfox_latest.zip"
+	}
+	manager := internal.NewSdkManager()
+	vfoxTempDir := manager.PathMeta.TempPath
+	tempFile = filepath.Join(vfoxTempDir, tempFile)
+	tempDir := filepath.Join(vfoxTempDir, "vfox_upgrade")
```
Comment (Chance-fyi) [inline cmd/commands/upgrade.go:140]:
```suggestion
	tempDir := filepath.Join(exeDir, "vfox_temp_upgrade")
	tempFile = filepath.Join(tempDir, tempFile)
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		return err
	}
```

**Review by Chance-fyi — APPROVED** [review]

File: cmd/commands/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	} else {
+		if err := os.Rename(tempExePath, exePath); err != nil {
+			return cli.Exit("Failed to replace executable: "+err.Error(), 1)
+		}
+		if err := os.Chmod(exePath, 0755); err != nil {
+			return cli.Exit("Failed to make executable: "+err.Error(), 1)
+		}
+	}
+
+	fmt.Printf("Updated to version: %s\n See the diff at: %s\n", latestVersion, diffURL)
```
Comment (Chance-fyi) [inline cmd/commands/upgrade.go:191]:
```suggestion
	fmt.Printf("Updated to version: %s\nSee the diff at: %s\n", latestVersion, diffURL)
```

**Review by aooohan — APPROVED** [review]

**Comment by aooohan** [comment]
@yanecc Hi, are you interested in joining `VersionFox`?

**Comment by yanecc** [comment]
> @yanecc Hi, are you interested in joining `VersionFox`?

Sure, It's my pleasure 😎

**Comment by aooohan** [comment]
> > @yanecc Hi, are you interested in joining `VersionFox`?
> 
> Sure, It's my pleasure 😎

I have invited you, please check your email and join the [Discord](https://discord.com/invite/85c8ptYgb7). ; )

**Pull request merged**